contracts/SeriesIndexer.sol

```diff
@@ -0,0 +1,792 @@
+
+    // Series Management
+    mapping(uint256 => Series) private seriesDetails;
+
+    // Series - Artist Relationship
+    mapping(uint256 => mapping(uint256 => bool)) private seriesArtistExist;
+
+    // Co-Artist Management
+    mapping(uint256 => mapping(uint256 => bool)) private seriesPendingCoArtist;
+    mapping(uint256 => mapping(uint256 => uint256)) private seriesPendingCoArtistIndex;
+    mapping(uint256 => mapping(uint256 => uint256)) private artistPendingCoArtistSeriesIndex;
+
+    // ============ Events ============
+
+    event ArtistAddressUpdated(
+        uint256 indexed artistID, 
+        address indexed oldAddress, 
+        address indexed newAddress
+    );
+    
+    event SeriesIndexed(
+        uint256 indexed seriesID,
+        uint256[] artistIDs,
+        string metadataURI,
+        string seriesContractTokenURI
+    );
+    
+    event SeriesUpdated(
+        uint256 indexed seriesID,
+        uint256[] artistIDs,
+        string metadataURI,
+        string seriesContractTokenURI
+    );
+    
+    event SeriesDeleted(uint256 indexed seriesID);
+    
+    event CoArtistProposed(
+        uint256 indexed seriesID, 
+        uint256 indexed proposedArtistID
+    );
+    
+    event CoArtistConfirmed(
+        uint256 indexed seriesID, 
+        uint256 indexed confirmedArtistID
+    );
+    
+    event CoArtistProposalCancelled(
+        uint256 indexed seriesID,
+        uint256 indexed proposerArtistID,
+        uint256 indexed cancelledArtistID
+    );
+
+    // ============ Modifiers ============
+
+    modifier onlyOwnerOrArtist(uint256 seriesID) {
+        if (msg.sender == owner()) {
+            // If owner is calling, check if at least one artist in this series has NOT revoked rights
+            if (!ownerCanModifySeries(seriesID)) {
+                revert OwnerRightsRevokedForThisSeries(seriesID);
+            }
+        } else {
+            // Otherwise, must be a valid artist in this series
+            _requireIsSeriesArtist(seriesID);
+        }
+        _;
+    }
+
+    modifier onlyArtist(uint256 seriesID) {
+        _requireIsSeriesArtist(seriesID);
+        _;
+    }
+
+    modifier seriesExists(uint256 seriesID) {
+        // If metadata is empty, we consider the series nonexistent
+        if (bytes(seriesDetails[seriesID].metadataURI).length == 0) {
+            revert SeriesDoesNotExistError(seriesID);
+        }
+        _;
+    }
+
+    // ============ Public/External Functions: Series Management ============
+
+    /**
+     * @dev Creates a new series
+     */
+    function addSeries(
+        address[] calldata artistAddrs,
+        string calldata metadataURI,
+        string calldata tokenIDsMapURI
+    ) external returns (uint256) {
+        return _createSeries(artistAddrs, metadataURI, tokenIDsMapURI);
+    }
+
+    /**
+     * @dev Batch creation of series
+     */
+    function batchAddSeries(
+        address[][] calldata artistsArray,
```

Comment:
This name could be confusing because of the word `array`, would just `artists` or `seriesArtists` be better?